File: problem-solving/adventofcode2020/day8.py
# https://adventofcode.com/2020/day/8
with open('./day8.txt', 'r') as f:
    input_list = [line for line in f.read().splitlines()]


class Command:
    def __init__(self, opr, arg):
        self.operation = opr
        self.argument = arg

    def convert_operation(self):
        if self.operation == "nop":
            self.operation = "jmp"
            return True
        elif self.operation == "jmp":
            self.operation = "nop"
            return True
        return False


def solve_first():
    visited = {}
    idx = 0
    accumulator = 0
    commands = parse_inputs_to_commands(input_list)
    while idx not in visited:
        visited[idx] = True
        opr, arg = commands[idx].operation, commands[idx].argument
        if opr == "nop":
            idx += 1
        elif opr == "acc":
            accumulator += arg
            idx += 1
        elif opr == "jmp":
            idx += arg
    return accumulator


def solve_second():
    commands = parse_inputs_to_commands(input_list)
    for cmd in commands:
        if not cmd.convert_operation():
            continue
        visited = {}
        idx = 0
        accumulator = 0
        is_continue = False
        while idx < len(commands):
            if idx not in visited:
                visited[idx] = True
            else:
                is_continue = True
                break
            visited[idx] = True
            opr, num = commands[idx].operation, commands[idx].argument
            if opr == "nop":
                idx += 1
            elif opr == "acc":
                accumulator += num
                idx += 1
            elif opr == "jmp":
                idx += num
        if is_continue:
            cmd.convert_operation()
            continue
        return accumulator


def parse_inputs_to_commands(input_list):
    commands = []
    for i in input_list:
        opr, arg = i.split()
        commands.append(Command(opr, int(arg)))
    return commands


if __name__ == "__main__":
    print(solve_first())
    print(solve_second())

File: problem-solving/adventofcode2020/day8.cpp
// https://adventofcode.com/2020/day/8
#include <algorithm>
#include <fstream>
#include <iostream>
#include <sstream>
#include <set>
#include <string>
#include <vector>
#include <unordered_map>
#include <iterator>

using namespace std;

class Command
{
public:
    string operation;
    int argument;

    Command(string opr, int arg)
    {
        operation = opr;
        argument = arg;
    }

    bool convertOperation()
    {
        if (operation == "nop")
        {
            operation = "jmp";
            return true;
        }
        else if (operation == "jmp")
        {
            operation = "nop";
            return true;
        }
        return false;
    }
};

vector<string> readInput(string filepath)
{
    vector<string> inputList;
    ifstream inputFile;
    inputFile.open(filepath);

    string line;
    while (getline(inputFile, line))
    {
        inputList.push_back(line);
    }

    inputFile.close();
    return inputList;
}

vector<Command> parseInputsToCommands(vector<string> &inputList)
{
    vector<Command> commands;
    for (int i = 0; i < inputList.size(); i++)
    {
        string input = inputList[i];
        stringstream ss(input);
        string operation;
        string argument;

        getline(ss, operation, ' ');
        getline(ss, argument, ' ');
        Command command(operation, stoi(argument));
        commands.push_back(command);
    }
    return commands;
}

int solvePart1(vector<Command> &commands)
{
    unordered_map<int, bool> visited;
    int idx = 0;
    int accumulator = 0;
    while (visited.count(idx) == 0)
    {
        visited[idx] = true;
        string opr = commands[idx].operation;
        int arg = commands[idx].argument;
        if (opr == "nop")
        {
            idx += 1;
        }
        else if (opr == "acc")
        {
            accumulator += arg;
            idx += 1;
        }
        else if (opr == "jmp")
        {
            idx += arg;
        }
    }
    return accumulator;
}

int solvePart2(vector<Command> &commands)
{
    for (int i = 0; i < commands.size(); i++)
    {
        if (!commands[i].convertOperation())
        {
            continue;
        }
        unordered_map<int, bool> visited;
        int idx = 0;
        int accumulator = 0;
        bool isContinue = false;
        while (idx < commands.size())
        {
            if (visited.count(idx) == 0)
            {
                visited[idx] = true;
            }
            else
            {
                isContinue = true;
                break;
            }
            visited[idx] = true;
            string opr = commands[idx].operation;
            int arg = commands[idx].argument;
            if (opr == "nop")
            {
                idx += 1;
            }
            else if (opr == "acc")
            {
                accumulator += arg;
                idx += 1;
            }
            else if (opr == "jmp")
            {
                idx += arg;
            }
        }
        if (isContinue)
        {
            commands[i].convertOperation();
            continue;
        }
        return accumulator;
    }
    return 0;
}

int main()
{
    vector<string> inputList = readInput("./day8.txt");
    vector<Command> commands = parseInputsToCommands(inputList);
    cout << solvePart1(commands) << endl;
    cout << solvePart2(commands) << endl;
}
